<a href="https://colab.research.google.com/github/CaroliCosas/Bootcamp_Data_Science/blob/main/03_pandas_1_ejercicios.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Ejercicios - pandas 1

In [ ]:
import pandas as pd

In [ ]:
# Versión

print(f"pandas=={pd.__version__}")

pandas==2.2.2


### Ejercicio 1:
- Lee el archivo **"Canada.xlsx"** y guardalo en la variable **df**

In [ ]:
df = pd.read_excel("Canada.xlsx")

In [ ]:
df.head(10)


Tipo                 Pais  Area                       Continente  \
0  Immigrants          Afghanistan   935                             Asia   
1  Immigrants              Albania   908                           Europe   
2  Immigrants              Algeria   903                           Africa   
3  Immigrants       American Samoa   909                          Oceania   
4  Immigrants              Andorra   908                           Europe   
5  Immigrants               Angola   903                           Africa   
6  Immigrants  Antigua and Barbuda   904  Latin America and the Caribbean   
7  Immigrants            Argentina   904  Latin America and the Caribbean   
8  Immigrants              Armenia   935                             Asia   
9  Immigrants            Australia   909                          Oceania   

   Region           Nombre de region      Tipo de region  1980  1981  1982  \
0    5501              Southern Asia  Developing regions    16    39    39   
1     925            Southern Europe   Developed regions     1     0     0   
2     912            Northern Africa  Developing regions    80    67    71   
3     957                  Polynesia  Developing regions     0     1     0   
4     925            Southern Europe   Developed regions     0     0     0   
5     911              Middle Africa  Developing regions     1     3     6   
6     915                  Caribbean  Developing regions     0     0     0   
7     931              South America  Developing regions   368   426   626   
8     922               Western Asia  Developing regions     0     0     0   
9     927  Australia and New Zealand   Developed regions   702   639   484   

   ...  2004  2005  2006  2007  2008  2009  2010  2011  2012  2013  
0  ...  2978  3436  3009  2652  2111  1746  1758  2203  2635  2004  
1  ...  1450  1223   856   702   560   716   561   539   620   603  
2  ...  3616  3626  4807  3623  4005  5393  4752  4325  3774  4331  
3  ...     0     0     1     0     0     0     0     0     0     0  
4  ...     0     0     1     1     0     0     0     0     1     1  
5  ...   268   295   184   106    76    62    61    39    70    45  
6  ...    14    24    32    15    32    38    27    37    51    25  
7  ...  1591  1153   847   620   540   467   459   278   263   282  
8  ...   147   224   218   198   205   267   252   236   258   207  
9  ...   930   909   875  1033  1018  1018   933   851   982  1121  

[10 rows x 41 columns]

### Ejercicio 2:
- Mostrar el tamaño de DataFrame.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Data columns (total 41 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Tipo              195 non-null    object
 1   Pais              195 non-null    object
 2   Area              195 non-null    int64 
 3   Continente        195 non-null    object
 4   Region            195 non-null    int64 
 5   Nombre de region  195 non-null    object
 6   Tipo de region    195 non-null    object
 7   1980              195 non-null    int64 
 8   1981              195 non-null    int64 
 9   1982              195 non-null    int64 
 10  1983              195 non-null    int64 
 11  1984              195 non-null    int64 
 12  1985              195 non-null    int64 
 13  1986              195 non-null    int64 
 14  1987              195 non-null    int64 
 15  1988              195 non-null    int64 
 16  1989              195 non-null    int64 
 17  1990            

In [ ]:
df.shape

(195, 41)

### Ejercicio 3:
- Mostrar las primeras 3 filas.

Tipo         Pais  Area Continente  Region Nombre de region  \
0  Immigrants  Afghanistan   935       Asia    5501    Southern Asia   
1  Immigrants      Albania   908     Europe     925  Southern Europe   
2  Immigrants      Algeria   903     Africa     912  Northern Africa   

       Tipo de region  1980  1981  1982  ...  2004  2005  2006  2007  2008  \
0  Developing regions    16    39    39  ...  2978  3436  3009  2652  2111   
1   Developed regions     1     0     0  ...  1450  1223   856   702   560   
2  Developing regions    80    67    71  ...  3616  3626  4807  3623  4005   

   2009  2010  2011  2012  2013  
0  1746  1758  2203  2635  2004  
1   716   561   539   620   603  
2  5393  4752  4325  3774  4331  

[3 rows x 41 columns]

In [ ]:
df.head(3)

Tipo         Pais  Area Continente  Region Nombre de region  \
0  Immigrants  Afghanistan   935       Asia    5501    Southern Asia   
1  Immigrants      Albania   908     Europe     925  Southern Europe   
2  Immigrants      Algeria   903     Africa     912  Northern Africa   

       Tipo de region  1980  1981  1982  ...  2004  2005  2006  2007  2008  \
0  Developing regions    16    39    39  ...  2978  3436  3009  2652  2111   
1   Developed regions     1     0     0  ...  1450  1223   856   702   560   
2  Developing regions    80    67    71  ...  3616  3626  4807  3623  4005   

   2009  2010  2011  2012  2013  
0  1746  1758  2203  2635  2004  
1   716   561   539   620   603  
2  5393  4752  4325  3774  4331  

[3 rows x 41 columns]

### Ejercicio 4:
- Mostrar las últimas 10 filas.

Tipo                                Pais  Area  \
185  Immigrants            United States of America   905   
186  Immigrants                             Uruguay   904   
187  Immigrants                          Uzbekistan   935   
188  Immigrants                             Vanuatu   909   
189  Immigrants  Venezuela (Bolivarian Republic of)   904   
190  Immigrants                            Viet Nam   935   
191  Immigrants                      Western Sahara   903   
192  Immigrants                               Yemen   935   
193  Immigrants                              Zambia   903   
194  Immigrants                            Zimbabwe   903   

                          Continente  Region    Nombre de region  \
185                 Northern America     905    Northern America   
186  Latin America and the Caribbean     931       South America   
187                             Asia    5500        Central Asia   
188                          Oceania     928           Melanesia   
189  Latin America and the Caribbean     931       South America   
190                             Asia     920  South-Eastern Asia   
191                           Africa     912     Northern Africa   
192                             Asia     922        Western Asia   
193                           Africa     910      Eastern Africa   
194                           Africa     910      Eastern Africa   

         Tipo de region  1980   1981  1982  ...  2004  2005  2006  2007  \
185   Developed regions  9378  10030  9074  ...  6990  8394  9613  9463   
186  Developing regions   128    132   146  ...   130   217   175   147   
187  Developing regions     0      0     0  ...   175   330   262   284   
188  Developing regions     0      0     0  ...     1     1     1     0   
189  Developing regions   103    117   174  ...  1224  1211  1192  1335   
190  Developing regions  1191   1829  2162  ...  1816  1852  3153  2574   
191  Developing regions     0      0     0  ...     0     0     1     0   
192  Developing regions     1      2     1  ...   124   161   140   122   
193  Developing regions    11     17    11  ...    56    91    77    71   
194  Developing regions    72    114   102  ...  1450   615   454   663   

      2008  2009  2010  2011  2012  2013  
185  10190  8995  8142  7676  7891  8501  
186    160    99    93    81    47    58  
187    215   288   289   162   235   167  
188      0     1     0     0     0     0  
189   1239  1353   998  1452  1373  1022  
190   1784  2171  1942  1723  1731  2112  
191      0     0     0     0     0     0  
192    133   128   211   160   174   217  
193     64    60   102    69    46    59  
194    611   508   494   434   437   407  

[10 rows x 41 columns]

In [ ]:
df.tail(10)

Tipo                                Pais  Area  \
185  Immigrants            United States of America   905   
186  Immigrants                             Uruguay   904   
187  Immigrants                          Uzbekistan   935   
188  Immigrants                             Vanuatu   909   
189  Immigrants  Venezuela (Bolivarian Republic of)   904   
190  Immigrants                            Viet Nam   935   
191  Immigrants                      Western Sahara   903   
192  Immigrants                               Yemen   935   
193  Immigrants                              Zambia   903   
194  Immigrants                            Zimbabwe   903   

                          Continente  Region    Nombre de region  \
185                 Northern America     905    Northern America   
186  Latin America and the Caribbean     931       South America   
187                             Asia    5500        Central Asia   
188                          Oceania     928           Melanesia   
189  Latin America and the Caribbean     931       South America   
190                             Asia     920  South-Eastern Asia   
191                           Africa     912     Northern Africa   
192                             Asia     922        Western Asia   
193                           Africa     910      Eastern Africa   
194                           Africa     910      Eastern Africa   

         Tipo de region  1980   1981  1982  ...  2004  2005  2006  2007  \
185   Developed regions  9378  10030  9074  ...  6990  8394  9613  9463   
186  Developing regions   128    132   146  ...   130   217   175   147   
187  Developing regions     0      0     0  ...   175   330   262   284   
188  Developing regions     0      0     0  ...     1     1     1     0   
189  Developing regions   103    117   174  ...  1224  1211  1192  1335   
190  Developing regions  1191   1829  2162  ...  1816  1852  3153  2574   
191  Developing regions     0      0     0  ...     0     0     1     0   
192  Developing regions     1      2     1  ...   124   161   140   122   
193  Developing regions    11     17    11  ...    56    91    77    71   
194  Developing regions    72    114   102  ...  1450   615   454   663   

      2008  2009  2010  2011  2012  2013  
185  10190  8995  8142  7676  7891  8501  
186    160    99    93    81    47    58  
187    215   288   289   162   235   167  
188      0     1     0     0     0     0  
189   1239  1353   998  1452  1373  1022  
190   1784  2171  1942  1723  1731  2112  
191      0     0     0     0     0     0  
192    133   128   211   160   174   217  
193     64    60   102    69    46    59  
194    611   508   494   434   437   407  

[10 rows x 41 columns]

### Ejercicio 5:
- Mostrar las estadisticas del DataFrame.

In [ ]:
df.describe()

Area       Region          1980          1981          1982  \
count  195.000000   195.000000    195.000000    195.000000    195.000000   
mean   912.764103  1249.015385    508.394872    566.989744    534.723077   
std     13.082835  1185.526885   1949.588546   2152.643752   1866.997511   
min    903.000000   905.000000      0.000000      0.000000      0.000000   
25%    903.000000   914.000000      0.000000      0.000000      0.000000   
50%    908.000000   922.000000     13.000000     10.000000     11.000000   
75%    922.000000   925.500000    251.500000    295.500000    275.000000   
max    935.000000  5501.000000  22045.000000  24796.000000  20620.000000   

               1983          1984         1985         1986          1987  \
count    195.000000    195.000000   195.000000   195.000000    195.000000   
mean     387.435897    376.497436   358.861538   441.271795    691.133333   
std     1204.333597   1198.246371  1079.309600  1225.576630   2109.205607   
min        0.000000      0.000000     0.000000     0.000000      0.000000   
25%        0.000000      0.000000     0.000000     0.500000      0.500000   
50%       12.000000     13.000000    17.000000    18.000000     26.000000   
75%      173.000000    181.000000   197.000000   254.000000    434.000000   
max    10015.000000  10170.000000  9564.000000  9470.000000  21337.000000   

       ...          2004          2005          2006          2007  \
count  ...    195.000000    195.000000    195.000000    195.000000   
mean   ...   1190.169231   1320.292308   1266.958974   1191.820513   
std    ...   3710.505369   4425.957828   3926.717747   3443.542409   
min    ...      0.000000      0.000000      0.000000      0.000000   
25%    ...     19.000000     28.500000     25.000000     31.000000   
50%    ...    191.000000    210.000000    218.000000    198.000000   
75%    ...    756.500000    832.000000    842.000000    899.000000   
max    ...  36619.000000  42584.000000  33848.000000  28742.000000   

               2008          2009          2010          2011          2012  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean    1246.394872   1275.733333   1420.287179   1262.533333   1313.958974   
std     3694.573544   3829.630424   4462.946328   4030.084313   4247.555161   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%       31.000000     36.000000     40.500000     37.500000     42.500000   
50%      205.000000    214.000000    211.000000    179.000000    233.000000   
75%      934.500000    888.000000    932.000000    772.000000    783.000000   
max    30037.000000  29622.000000  38617.000000  36765.000000  34315.000000   

               2013  
count    195.000000  
mean    1320.702564  
std     4237.951988  
min        0.000000  
25%       45.000000  
50%      213.000000  
75%      796.000000  
max    34129.000000  

[8 rows x 36 columns]

### Ejercicio 6:
- Mostrar la información del DataFrame.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Data columns (total 41 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Tipo              195 non-null    object
 1   Pais              195 non-null    object
 2   Area              195 non-null    int64 
 3   Continente        195 non-null    object
 4   Region            195 non-null    int64 
 5   Nombre de region  195 non-null    object
 6   Tipo de region    195 non-null    object
 7   1980              195 non-null    int64 
 8   1981              195 non-null    int64 
 9   1982              195 non-null    int64 
 10  1983              195 non-null    int64 
 11  1984              195 non-null    int64 
 12  1985              195 non-null    int64 
 13  1986              195 non-null    int64 
 14  1987              195 non-null    int64 
 15  1988              195 non-null    int64 
 16  1989              195 non-null    int64 
 17  1990            

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Data columns (total 41 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Tipo              195 non-null    object
 1   Pais              195 non-null    object
 2   Area              195 non-null    int64 
 3   Continente        195 non-null    object
 4   Region            195 non-null    int64 
 5   Nombre de region  195 non-null    object
 6   Tipo de region    195 non-null    object
 7   1980              195 non-null    int64 
 8   1981              195 non-null    int64 
 9   1982              195 non-null    int64 
 10  1983              195 non-null    int64 
 11  1984              195 non-null    int64 
 12  1985              195 non-null    int64 
 13  1986              195 non-null    int64 
 14  1987              195 non-null    int64 
 15  1988              195 non-null    int64 
 16  1989              195 non-null    int64 
 17  1990            

### Ejercicio 7:
- Mostar las columnas del DataFrame en una lista.

Index([            'Tipo',             'Pais',             'Area',
             'Continente',           'Region', 'Nombre de region',
         'Tipo de region',               1980,               1981,
                     1982,               1983,               1984,
                     1985,               1986,               1987,
                     1988,               1989,               1990,
                     1991,               1992,               1993,
                     1994,               1995,               1996,
                     1997,               1998,               1999,
                     2000,               2001,               2002,
                     2003,               2004,               2005,
                     2006,               2007,               2008,
                     2009,               2010,               2011,
                     2012,               2013],
      dtype='object')

In [ ]:
df.columns.tolist()

['Tipo',
 'Pais',
 'Area',
 'Continente',
 'Region',
 'Nombre de region',
 'Tipo de region',
 1980,
 1981,
 1982,
 1983,
 1984,
 1985,
 1986,
 1987,
 1988,
 1989,
 1990,
 1991,
 1992,
 1993,
 1994,
 1995,
 1996,
 1997,
 1998,
 1999,
 2000,
 2001,
 2002,
 2003,
 2004,
 2005,
 2006,
 2007,
 2008,
 2009,
 2010,
 2011,
 2012,
 2013]

### Ejercicio 8:
- Eliminar las columnas: _**Tipo**_, _**Area**_ y _**Region**_.

In [ ]:
# df.drop(["Type", "Area", "Region"], axis=1, inplace=True)


KeyError: "['Type'] not found in axis"

### Ejercicio 9:
- Obtener los valores únicos de la columna _**Continente**_ en una lista.

array(['Asia', 'Europe', 'Africa', 'Oceania',
       'Latin America and the Caribbean', 'Northern America'],
      dtype=object)

In [ ]:
df.unique(Continente)

AttributeError: 'DataFrame' object has no attribute 'unique'

### Ejercicio 10:
- Obtener el valor minimo del año 2000.

0

### Ejercicio 11:
- Obtener los valores máximos de 1991-2000.

1991    23380
1992    34123
1993    33720
1994    39231
1995    30145
1996    29322
1997    22965
1998    21049
1999    30069
2000    35529
dtype: int64

### Ejercicio 12:
- Obtener una lista con la media de migración de todos los años.

[1064.148717948718,
 1136.8564102564103,
 1138.7128205128206,
 993.1538461538462,
 962.625641025641,
 1026.076923076923,
 989.1538461538462,
 824.2410256410257,
 922.1435897435897,
 1111.3435897435897,
 1244.323076923077,
 1144.1589743589743,
 1114.3435897435897,
 1190.1692307692308,
 1320.2923076923078,
 1266.9589743589743,
 1191.820512820513,
 1246.3948717948717,
 1275.7333333333333,
 1420.2871794871794,
 1262.5333333333333,
 1313.9589743589743,
 1320.702564102564]

### Ejercicio 13:
- Obtener el país con más migración en el año 2010.

Pais Continente    Nombre de region      Tipo de region  1980  \
136  Philippines       Asia  South-Eastern Asia  Developing regions  6051   

     1981  1982  1983  1984  1985  ...   2004   2005   2006   2007   2008  \
136  5921  5249  4562  3801  3150  ...  14004  18139  18400  19837  24887   

      2009   2010   2011   2012   2013  
136  28573  38617  36765  34315  29544  

[1 rows x 38 columns]

### Ejercicio 14
- Obtener el país con menor migración en el año 2005.

Pais Continente Nombre de region      Tipo de region  1980  \
109  Marshall Islands    Oceania       Micronesia  Developing regions     0   

     1981  1982  1983  1984  1985  ...  2004  2005  2006  2007  2008  2009  \
109     0     0     0     0     0  ...     0     0     0     2     0     0   

     2010  2011  2012  2013  
109     0     0     0     0  

[1 rows x 38 columns]

### Ejercicio 15:
- Encuentra el año en que más migración hubo de España a Canada.

161
2013  331

### Ejercicio 16:
- Encuentra el año en que menor migración hubo de España a Canada.

161
1995   78

### Ejercicio 17:
- Ordena en una lista los años de migración en España de menor a mayor de acuerdo a la información de cada año.

[1995,
 2001,
 1999,
 1994,
 2002,
 1998,
 1997,
 2003,
 2000,
 1985,
 1996,
 1986,
 1990,
 2006,
 2005,
 1992,
 2004,
 1983,
 2007,
 1988,
 1984,
 1993,
 1991,
 1989,
 2008,
 2010,
 1987,
 2011,
 2009,
 1980,
 1982,
 2012,
 1981,
 2013]

### Ejercicio 18:
- Crea una lista con todos los países.

array(['Afghanistan', 'Albania', 'Algeria', 'American Samoa', 'Andorra',
       'Angola', 'Antigua and Barbuda', 'Argentina', 'Armenia',
       'Australia', 'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain',
       'Bangladesh', 'Barbados', 'Belarus', 'Belgium', 'Belize', 'Benin',
       'Bhutan', 'Bolivia (Plurinational State of)',
       'Bosnia and Herzegovina', 'Botswana', 'Brazil',
       'Brunei Darussalam', 'Bulgaria', 'Burkina Faso', 'Burundi',
       'Cabo Verde', 'Cambodia', 'Cameroon', 'Canada',
       'Central African Republic', 'Chad', 'Chile', 'China',
       'China, Hong Kong Special Administrative Region',
       'China, Macao Special Administrative Region', 'Colombia',
       'Comoros', 'Congo', 'Costa Rica', "Côte d'Ivoire", 'Croatia',
       'Cuba', 'Cyprus', 'Czech Republic',
       "Democratic People's Republic of Korea",
       'Democratic Republic of the Congo', 'Denmark', 'Djibouti',
       'Dominica', 'Dominican Republic', 'Ecuador', 'Egypt',
       'El Salvador', 

### Ejercicio 19:
- Encuentra el número de paises por continente y agrega los nombres a un diccionario.
    - Ejemplo: **{"Asía" : 20, "Africa" : 10, "Europa" : 5}**

{'Africa': 54,
 'Asia': 49,
 'Europe': 43,
 'Latin America and the Caribbean': 33,
 'Oceania': 14,
 'Northern America': 2}

### Ejercicio 20:
- Crea un DataFrame a partir del diccionario anterior.

Continente  Count
0                           Africa     54
1                             Asia     49
2                           Europe     43
3  Latin America and the Caribbean     33
4                          Oceania     14
5                 Northern America      2

### Ejercicio 21:
- Repite los últimos dos ejercicios tomando en cuenta que america latina y america del norte son un solo continente.

{'Africa': 54,
 'Asia': 49,
 'Europe': 43,
 'Latin America and the Caribbean': 33,
 'Oceania': 14,
 'Northern America': 2}

{'Africa': 54,
 'Asia': 49,
 'Europe': 43,
 'Latin America and the Caribbean': 33,
 'Oceania': 14,
 'Northern America': 2,
 'America': 35}

Continente  Count
0     Africa     54
1       Asia     49
2     Europe     43
4    Oceania     14
6    America     35

### Ejercicio 22:
- Crea un DataFrame por continente y en cada DataFrame apareceran los paises de dicho contiente.
- Muestra las 3 primeras filas de cada DataFrame.

Pais Continente Nombre de region      Tipo de region  1980  1981  \
0   Afghanistan       Asia    Southern Asia  Developing regions    16    39   
8       Armenia       Asia     Western Asia  Developing regions     0     0   
11   Azerbaijan       Asia     Western Asia  Developing regions     0     0   

    1982  1983  1984  1985  ...  2004  2005  2006  2007  2008  2009  2010  \
0     39    47    71   340  ...  2978  3436  3009  2652  2111  1746  1758   
8      0     0     0     0  ...   147   224   218   198   205   267   252   
11     0     0     0     0  ...   230   359   236   203   125   165   209   

    2011  2012  2013  
0   2203  2635  2004  
8    236   258   207  
11   138   161    57  

[3 rows x 38 columns]

Pais Continente Nombre de region     Tipo de region  1980  1981  1982  \
1   Albania     Europe  Southern Europe  Developed regions     1     0     0   
4   Andorra     Europe  Southern Europe  Developed regions     0     0     0   
10  Austria     Europe   Western Europe  Developed regions   234   238   201   

    1983  1984  1985  ...  2004  2005  2006  2007  2008  2009  2010  2011  \
1      0     0     0  ...  1450  1223   856   702   560   716   561   539   
4      0     0     0  ...     0     0     1     1     0     0     0     0   
10   117   127   165  ...   118   122   102    98    91   109   124    84   

    2012  2013  
1    620   603  
4      1     1  
10   102    90  

[3 rows x 38 columns]

Pais Continente Nombre de region      Tipo de region  1980  1981  1982  \
2   Algeria     Africa  Northern Africa  Developing regions    80    67    71   
5    Angola     Africa    Middle Africa  Developing regions     1     3     6   
19    Benin     Africa   Western Africa  Developing regions     2     5     4   

    1983  1984  1985  ...  2004  2005  2006  2007  2008  2009  2010  2011  \
2     69    63    44  ...  3616  3626  4807  3623  4005  5393  4752  4325   
5      6     4     3  ...   268   295   184   106    76    62    61    39   
19     3     4     3  ...   112    95   116   183   205   238   290   284   

    2012  2013  
2   3774  4331  
5     70    45  
19   391   397  

[3 rows x 38 columns]

Pais Continente           Nombre de region      Tipo de region  \
3   American Samoa    Oceania                  Polynesia  Developing regions   
9        Australia    Oceania  Australia and New Zealand   Developed regions   
61            Fiji    Oceania                  Melanesia  Developing regions   

    1980  1981  1982  1983  1984  1985  ...  2004  2005  2006  2007  2008  \
3      0     1     0     0     0     0  ...     0     0     1     0     0   
9    702   639   484   317   317   319  ...   930   909   875  1033  1018   
61   632   705   814   554   389   446  ...   495   298   277   304   321   

    2009  2010  2011  2012  2013  
3      0     0     0     0     0  
9   1018   933   851   982  1121  
61   311   398   316   270   228  

[3 rows x 38 columns]

Pais                       Continente Nombre de region  \
6   Antigua and Barbuda  Latin America and the Caribbean        Caribbean   
7             Argentina  Latin America and the Caribbean    South America   
12              Bahamas  Latin America and the Caribbean        Caribbean   

        Tipo de region  1980  1981  1982  1983  1984  1985  ...  2004  2005  \
6   Developing regions     0     0     0     0    42    52  ...    14    24   
7   Developing regions   368   426   626   241   237   196  ...  1591  1153   
12  Developing regions    26    23    38    12    21    28  ...    13    14   

    2006  2007  2008  2009  2010  2011  2012  2013  
6     32    15    32    38    27    37    51    25  
7    847   620   540   467   459   278   263   282  
12    18    23    30    27    25    40    60    55  

[3 rows x 38 columns]

Pais        Continente  Nombre de region  \
32                     Canada  Northern America  Northern America   
185  United States of America  Northern America  Northern America   

        Tipo de region  1980   1981  1982  1983  1984  1985  ...  2004  2005  \
32   Developed regions     0      0     0     0     0     0  ...     0     0   
185  Developed regions  9378  10030  9074  7100  6661  6543  ...  6990  8394   

     2006  2007   2008  2009  2010  2011  2012  2013  
32      0     0      0     0     0     1     1     2  
185  9613  9463  10190  8995  8142  7676  7891  8501  

[2 rows x 38 columns]

Pais                       Continente  \
0                 Afghanistan                             Asia   
8                     Armenia                             Asia   
11                 Azerbaijan                             Asia   
1                     Albania                           Europe   
4                     Andorra                           Europe   
10                    Austria                           Europe   
2                     Algeria                           Africa   
5                      Angola                           Africa   
19                      Benin                           Africa   
3              American Samoa                          Oceania   
9                   Australia                          Oceania   
61                       Fiji                          Oceania   
6         Antigua and Barbuda  Latin America and the Caribbean   
7                   Argentina  Latin America and the Caribbean   
12                    Bahamas  Latin America and the Caribbean   
32                     Canada                 Northern America   
185  United States of America                 Northern America   

              Nombre de region      Tipo de region  1980   1981  1982  1983  \
0                Southern Asia  Developing regions    16     39    39    47   
8                 Western Asia  Developing regions     0      0     0     0   
11                Western Asia  Developing regions     0      0     0     0   
1              Southern Europe   Developed regions     1      0     0     0   
4              Southern Europe   Developed regions     0      0     0     0   
10              Western Europe   Developed regions   234    238   201   117   
2              Northern Africa  Developing regions    80     67    71    69   
5                Middle Africa  Developing regions     1      3     6     6   
19              Western Africa  Developing regions     2      5     4     3   
3                    Polynesia  Developing regions     0      1     0     0   
9    Australia and New Zealand   Developed regions   702    639   484   317   
61                   Melanesia  Developing regions   632    705   814   554   
6                    Caribbean  Developing regions     0      0     0     0   
7                South America  Developing regions   368    426   626   241   
12                   Caribbean  Developing regions    26     23    38    12   
32            Northern America   Developed regions     0      0     0     0   
185           Northern America   Developed regions  9378  10030  9074  7100   

     1984  1985  ...  2004  2005  2006  2007   2008  2009  2010  2011  2012  \
0      71   340  ...  2978  3436  3009  2652   2111  1746  1758  2203  2635   
8       0     0  ...   147   224   218   198    205   267   252   236   258   
11      0     0  ...   230   359   236   203    125   165   209   138   161   
1       0     0  ...  1450  1223   856   702    560   716   561   539   620   
4       0     0  ...     0     0     1     1      0     0     0     0     1   
10    127   165  ...   118   122   102    98     91   109   124    84   102   
2      63    44  ...  3616  3626  4807  3623   4005  5393  4752  4325  3774   
5       4     3  ...   268   295   184   106     76    62    61    39    70   
19      4     3  ...   112    95   116   183    205   238   290   284   391   
3       0     0  ...     0     0     1     0      0     0     0     0     0   
9     317   319  ...   930   909   875  1033   1018  1018   933   851   982   
61    389   446  ...   495   298   277   304    321   311   398   316   270   
6      42    52  ...    14    24    32    15     32    38    27    37    51   
7     237   196  ...  1591  1153   847   620    540   467   459   278   263   
12     21    28  ...    13    14    18    23     30    27    25    40    60   
32      0     0  ...     0     0     0     0      0     0     0     1     1   
185  6661  6543  ...  6990  8394  9613  9463  10190  8995  8142  7676 

### Ejercicio 23:
- Obtener los valores maximos de migración de cada año y crea un DataFrame con los años y el número de migración máximo de cada año.

|     |Año |Numero de Migrantes|
|-----|----|-------------------|
|**0**|1980|22045              |
|**1**|1981|24796              |
|**2**|1982|20620              |
|**3**|1983|10015              |
|**4**|1984|10170              |

Año  Numero de Migrantes
0  1980                22045
1  1981                24796
2  1982                20620
3  1983                10015
4  1984                10170

Año  Numero de Migrantes
0   1980                22045
1   1981                24796
2   1982                20620
3   1983                10015
4   1984                10170
5   1985                 9564
6   1986                 9470
7   1987                21337
8   1988                27359
9   1989                23795
10  1990                31668
11  1991                23380
12  1992                34123
13  1993                33720
14  1994                39231
15  1995                30145
16  1996                29322
17  1997                22965
18  1998                21049
19  1999                30069
20  2000                35529
21  2001                36434
22  2002                31961
23  2003                36439
24  2004                36619
25  2005                42584
26  2006                33848
27  2007                28742
28  2008                30037
29  2009                29622
30  2010                38617
31  2011                36765
32  2012                34315
33  2013                34129

### Ejercicio 24:
- Ordenar por número mayor de migración a menor el DataFrame anterior.

Año  Numero de Migrantes
25  2005                42584
14  1994                39231
30  2010                38617
31  2011                36765
24  2004                36619
23  2003                36439
21  2001                36434
20  2000                35529
32  2012                34315
33  2013                34129
12  1992                34123
26  2006                33848
13  1993                33720
22  2002                31961
10  1990                31668
15  1995                30145
19  1999                30069
28  2008                30037
29  2009                29622
16  1996                29322
27  2007                28742
8   1988                27359
1   1981                24796
9   1989                23795
11  1991                23380
17  1997                22965
0   1980                22045
7   1987                21337
18  1998                21049
2   1982                20620
4   1984                10170
3   1983                10015
5   1985                 9564
6   1986                 9470

### Ejercicio 25:
- Obtener el promedio y la mediana de cada año de migración y agregar al dataframe anterior.


|     |Año |Numero de Migrantes|promedio  |mediana|
|-----|----|-------------------|----------|-------|
|**0**|1980|22045              |508.394872|13.0   |
|**1**|1981|24796              |566.989744|10.0   |
|**2**|1982|20620              |534.723077|11.0   |
|**3**|1983|10015              |387.435897|12.0   |
|**4**|1984|10170              |376.497436|13.0   |

Año  Numero de Migrantes    promedio  mediana
0  1980                22045  508.394872     13.0
1  1981                24796  566.989744     10.0
2  1982                20620  534.723077     11.0
3  1983                10015  387.435897     12.0
4  1984                10170  376.497436     13.0

### Ejercicio 26:
- Encontrar el año en que hubo más migración por continente y crear un dataframe con nombre de continente, el año de máxima migración y cantidad de migración.

|     |Continente|Año |Migraciones|
|-----|----------|----|-----------|
|**0**|Asia      |2010|163845     |
|**1**|Europe    |1990|68301      |
|**2**|Africa    |2010|40892      |
|**3**|Oceania   |1992|2871       |
|**4**|America   |1992|44699      |

array(['Asia', 'Europe', 'Africa', 'Oceania', 'America'], dtype=object)

In [ ]:
df_26

Continente   Año  Migraciones
0       Asia  2010       163845
1     Europe  1990        68301
2     Africa  2010        40892
3    Oceania  1992         2871
4    America  1992        44699

### Ejercicio 27:
- Encontrar los 10 paises con más migración entre el 2008 y 2010.

|       |Pais                                             |2008 |2009 |2010|migracion_2010|
|-------|-------------------------------------------------|-----|-----|-----|--------------|
|**136**|Philippines                                      |24887|28573|38617|         92077|
|**79** |India                                            |28261|29456|34235|         91952|
|**36** |China                                            |30037|29622|30391|         90050|
|**185**|United States of America                         |10190|8995 |8142 |         27327|
|**183**|United Kingdom of Great Britain and Northern I...|8979 |8876 |8724 |         26579|
|**130**|Pakistan                                         |8994 |7217 |6811 |         23022|
|**81** |Iran (Islamic Republic of)                       |6475 |6580 |7477 |         20532|
|**140**|Republic of Korea                                |7294 |5874 |5537 |         18705|
|**116**|Morocco                                          |4226 |5532 |6242 |         16000|
|**39** |Colombia                                         |5452 |4652 |5218 |         15322|

Pais Continente Nombre de region      Tipo de region  1980  1981  \
0  Afghanistan       Asia    Southern Asia  Developing regions    16    39   
1      Albania     Europe  Southern Europe   Developed regions     1     0   
2      Algeria     Africa  Northern Africa  Developing regions    80    67   

   1982  1983  1984  1985  ...  2005  2006  2007  2008  2009  2010  2011  \
0    39    47    71   340  ...  3436  3009  2652  2111  1746  1758  2203   
1     0     0     0     0  ...  1223   856   702   560   716   561   539   
2    71    69    63    44  ...  3626  4807  3623  4005  5393  4752  4325   

   2012  2013  migracion_2010  
0  2635  2004            5615  
1   620   603            1837  
2  3774  4331           14150  

[3 rows x 39 columns]

Pais   2008   2009   2010  \
136                                        Philippines  24887  28573  38617   
79                                               India  28261  29456  34235   
36                                               China  30037  29622  30391   
185                           United States of America  10190   8995   8142   
183  United Kingdom of Great Britain and Northern I...   8979   8876   8724   
130                                           Pakistan   8994   7217   6811   
81                          Iran (Islamic Republic of)   6475   6580   7477   
140                                  Republic of Korea   7294   5874   5537   
116                                            Morocco   4226   5532   6242   
39                                            Colombia   5452   4652   5218   

     migracion_2010  
136           92077  
79            91952  
36            90050  
185           27327  
183           26579  
130           23022  
81            20532  
140           18705  
116           16000  
39            15322

### Ejercicio 28:
- Agregar una columna al **DataFrame inicial** con el número total de migración de 1980 al 2013 por pais.

Pais Continente Nombre de region      Tipo de region  1980  1981  \
0     Afghanistan       Asia    Southern Asia  Developing regions    16    39   
1         Albania     Europe  Southern Europe   Developed regions     1     0   
2         Algeria     Africa  Northern Africa  Developing regions    80    67   
3  American Samoa    Oceania        Polynesia  Developing regions     0     1   
4         Andorra     Europe  Southern Europe   Developed regions     0     0   

   1982  1983  1984  1985  ...  2006  2007  2008  2009  2010  2011  2012  \
0    39    47    71   340  ...  3009  2652  2111  1746  1758  2203  2635   
1     0     0     0     0  ...   856   702   560   716   561   539   620   
2    71    69    63    44  ...  4807  3623  4005  5393  4752  4325  3774   
3     0     0     0     0  ...     1     0     0     0     0     0     0   
4     0     0     0     0  ...     1     1     0     0     0     0     1   

   2013  migracion_2010  Total Migracion  
0  2004            5615            12457  
1   603            1837             3599  
2  4331           14150            26580  
3     0               0                0  
4     1               0                2  

[5 rows x 40 columns]

### Ejercicio 29:
- Encuentra el país de cada continente con más migración y crea un DataFrame con estos paises.

Pais Continente  \
136                                        Philippines       Asia   
183  United Kingdom of Great Britain and Northern I...     Europe   
116                                            Morocco     Africa   
9                                            Australia    Oceania   
185                           United States of America    America   

              Nombre de region      Tipo de region   1980   1981   1982  \
136         South-Eastern Asia  Developing regions   6051   5921   5249   
183            Northern Europe   Developed regions  22045  24796  20620   
116            Northern Africa  Developing regions    325    471    447   
9    Australia and New Zealand   Developed regions    702    639    484   
185           Northern America   Developed regions   9378  10030   9074   

      1983   1984  1985  ...   2006   2007   2008   2009   2010   2011   2012  \
136   4562   3801  3150  ...  18400  19837  24887  28573  38617  36765  34315   
183  10015  10170  9564  ...   7140   8216   8979   8876   8724   6204   6195   
116    335    248   328  ...   3322   4021   4226   5532   6242   4399   3878   
9      317    317   319  ...    875   1033   1018   1018    933    851    982   
185   7100   6661  6543  ...   9613   9463  10190   8995   8142   7676   7891   

      2013  migracion_2010  Total Migracion  
136  29544           92077           192701  
183   5827           26579            44805  
116   3261           16000            27538  
9     1121            2969             5923  
185   8501           27327            51395  

[5 rows x 40 columns]

### Ejercicio 30:
- Encontrar los 3 paises con más migración de cada continente y agregarlos en un DataFrame.

|       |Continente|Pais                                             |Total Migracion|
|-------|----------|-------------------------------------------------|---------------|
|**79**	|Asia	   |India                                            |691904         |
|**36**	|Asia	   |China                                            |659962         |
|**136**|Asia	   |Philippines                                      |511391         |
|**183**|Europa	   |United Kingdom of Great Britain and Northern I...|551500         |
|**137**|Europa	   |Poland                                           |139241         |
|**63**	|Europa	   |France                                           |109091         |
|**55**	|Africa	   |Egypt                                            |72745          |
|**116**|Africa	   |Morocco                                          |72267          |
|**2**	|Africa	   |Algeria                                          |69439          |
|**9**	|Oceania   |Australia                                        |23829          |
|**61**	|Oceania   |Fiji                                             |20609          |
|**124**|Oceania   |New Zealand                                      |10386          |
|**185**|America   |United States of America                         |241122         |
|**86**	|America   |Jamaica                                          |106431         |
|**75**	|America   |Haiti                                            |84813          |

Pais  Total Migracion
136                                        Philippines           192701
36                                               China           185705
79                                               India           183481
183  United Kingdom of Great Britain and Northern I...            44805
63                                              France            30212
67                                             Germany            15648
116                                            Morocco            27538
55                                               Egypt            27208
2                                              Algeria            26580
9                                            Australia             5923
124                                        New Zealand             2697
61                                                Fiji             1844
185                           United States of America            51395
39                                            Colombia            27060
75                                               Haiti            25838

### Ejercicio 31:
- Cambiar los nombres de cada continente a español.

array(['Asia', 'Europa', 'Africa', 'Oceania', 'America'], dtype=object)

#### Extra: Obtener un Dataframe que te parezca relevante a partir de la aplicación de filtros. ¡Se creativo!

In [ ]:
################################################################################################################################